cranelift/codegen/src/isa/x86/enc_tables.rs

```diff
@@ -1513,6 +1513,47 @@ fn convert_ishl(
     }
 }
 
+/// Convert an imul.i64x2 to a valid code sequence on x86, first with AVX512 and then with SSE2.
+fn convert_i64x2_imul(
+    inst: ir::Inst,
+    func: &mut ir::Function,
+    _cfg: &mut ControlFlowGraph,
+    isa: &dyn TargetIsa,
+) {
+    let mut pos = FuncCursor::new(func).at_inst(inst);
+    pos.use_srcloc(inst);
+
+    if let ir::InstructionData::Binary {
+        opcode: ir::Opcode::Imul,
+        args: [arg0, arg1],
+    } = pos.func.dfg[inst]
+    {
+        let ty = pos.func.dfg.ctrl_typevar(inst);
+        if ty == I64X2 {
+            let x86_isa = isa
+                .as_any()
+                .downcast_ref::<isa::x86::Isa>()
+                .expect("the target ISA must be x86 at this point");
+            if x86_isa.isa_flags.use_avx512dq_simd() || x86_isa.isa_flags.use_avx512vl_simd() {
+                // If we have certain AVX512 features, we can lower this instruction simply.
+                pos.func.dfg.replace(inst).x86_pmullq(arg0, arg1);
+            } else {
+                // Otherwise, we default to a very lengthy SSE2-compatible sequence.
+                let high0 = pos.ins().ushr_imm(arg0, 32);
+                let mul0 = pos.ins().x86_pmuludq(high0, arg1);
+                let high1 = pos.ins().ushr_imm(arg1, 32);
+                let mul1 = pos.ins().x86_pmuludq(high1, arg0);
+                let addhigh = pos.ins().iadd(mul0, mul1);
+                let high = pos.ins().ishl_imm(addhigh, 32);
+                let low = pos.ins().x86_pmuludq(arg0, arg1);
+                pos.func.dfg.replace(inst).iadd(low, high);
+            }
+        } else {
+            // Other imul variants should be encodable.
```

Comment:
This doesn't look quite right to me but it may be possible I'm misunderstanding because I don't know the Cranelift IR.  The result of addhigh looks right, it looks like it has the results for the high dwords of the result qwords in its low two dwords, ie the upper two dwords are garbage.  But the value of 'high' isn't right I think, at least I can't make either psllq or pslldq work here - some kind of shuffle might work.  The computation of low computes the two low dwords of the result qwords correctly in the two low dwords, but simply adding them in won't work even if high was correct, they again need to be shuffled into their correct places.  Instead of shift and add it looks like an unpack might work (this is untested):
```
  // lhsDest = D C B A
  // rhs = H G F E
  mov(lhsDest, high0);       // high0 = D C B A
  psrlq(32, high0);          // high0 = 0 D 0 B
  pmuludq(rhs, high0);       // high0 = 0 D CG BE
  mov(rhs, high1);           // high1 = H G F E
  psrlq(32, high1);          // high1 = 0 H 0 F
  pmuludq(lhsDest, high1);   // high1 = 0 H CH AF
  paddq(high0, high1);       // high1 = 0 D+H CG+CH BE+AF
  pmuludq(rhs, lhsDest);     // lhsDest = D C CG AE
  punpckldq(high1, lhsDest); // lhsDest = CG+CH CG BE+AF AE
```

Reply:
Benjamin asked me to take a look.  Initially I had a comment here where I indicated I thought this was wrong, but I see I misunderstood something, and I now think this looks right to me.